src/etl/ProbingETLSource.cpp

```diff
@@ -0,0 +1,187 @@
+    return currentSrc_->toString();
+}
+
+bool
+ProbingETLSource::loadInitialLedger(
+    std::uint32_t ledgerSequence,
+    std::uint32_t numMarkers,
+    bool cacheOnly)
+{
+    if (!currentSrc_)
+        return false;
+    return currentSrc_->loadInitialLedger(
+        ledgerSequence, numMarkers, cacheOnly);
+}
+
+std::pair<grpc::Status, org::xrpl::rpc::v1::GetLedgerResponse>
+ProbingETLSource::fetchLedger(
+    uint32_t ledgerSequence,
+    bool getObjects,
+    bool getObjectNeighbors)
+{
+    if (!currentSrc_)
+        return {};
+    return currentSrc_->fetchLedger(
+        ledgerSequence, getObjects, getObjectNeighbors);
+}
+
+std::optional<boost::json::object>
+ProbingETLSource::forwardToRippled(
+    boost::json::object const& request,
+    std::string const& clientIp,
+    boost::asio::yield_context& yield) const
+{
+    assert(currentSrc_);  // expected to be connected
+    return currentSrc_->forwardToRippled(request, clientIp, yield);
+}
+
+std::optional<boost::json::object>
+ProbingETLSource::requestFromRippled(
+    boost::json::object const& request,
+    std::string const& clientIp,
+    boost::asio::yield_context& yield) const
+{
+    assert(currentSrc_);  // expected to be connected
+    return currentSrc_->requestFromRippled(request, clientIp, yield);
+}
+
+ETLSourceHooks
+ProbingETLSource::make_SSLHooks() noexcept
+{
+    return {
+        std::ref(mtx_),
+        // onConnected
+        [this](auto ec) {
+            if (currentSrc_)
+                return ETLSourceHooks::Action::STOP;
+
+            if (!ec)
+            {
+                plainSrc_->pause();
+                currentSrc_ = sslSrc_;
+                BOOST_LOG_TRIVIAL(info) << "Selected WSS as the main source: "
+                                        << currentSrc_->toString();
+            }
+            return ETLSourceHooks::Action::PROCEED;
+        },
+        // onDisconnected
+        [this](auto ec) {
+            if (currentSrc_)
+            {
+                currentSrc_ = nullptr;
+                plainSrc_->resume();
+            }
+            return ETLSourceHooks::Action::STOP;
+        }};
+}
+
+ETLSourceHooks
+ProbingETLSource::make_PlainHooks() noexcept
+{
+    return {
+        std::ref(mtx_),
+        // onConnected
+        [this](auto ec) {
+            if (currentSrc_)
+                return ETLSourceHooks::Action::STOP;
+
+            if (!ec)
+            {
+                sslSrc_->pause();
+                currentSrc_ = plainSrc_;
+                BOOST_LOG_TRIVIAL(info)
+                    << "Selected Plain WS as the main source: "
+                    << currentSrc_->toString();
+            }
+            return ETLSourceHooks::Action::PROCEED;
+        },
+        // onDisconnected
+        [this](auto ec) {
+            if (currentSrc_)
+            {
+                currentSrc_ = nullptr;
+                sslSrc_->resume();
+            }
+            return ETLSourceHooks::Action::STOP;
+        }};
+}
```

Comment:
I quite like your lambda-based implementation, might try to do this.

In general tho, you are guided by the `DRY principle` (i assume). It's good to remember that `DRY` is not about "never repeat code" but rather about "what will be easier when I need to change this in the future". 
So if the logic will change (and it likely will eventually) then your approach actually will require more changes to the code than two separate functions with _sortof_ copied code. It's always a trade off..

Reply:
I tried to implement this in a few different ways but since hooks rely on both `ssl` and `plain` sources and are injected in the constructor in the initialization list of `probing source` there is currently not really a good non-hacky way to do this. 
I think it's not worth to do the refactor at this moment, especially because after v1 we wanted to refactor the `etl` subsystem anyways. I'd leave this as is for now except for the mutex which is now captured as you suggested.